hardware_interface/include/hardware_interface/component_parser.hpp

```diff
@@ -0,0 +1,113 @@
+#include "hardware_interface/visibility_control.h"
+
+namespace hardware_interface
+{
+
+/**
+  * \brief Search XML snippet from URDF for informations about a control component.
+  *
+  * \param urdf string with robot's URDF
+  * \return vector filled with informations about robot's control ressources
+  * \throws std::runtime_error if a robot attribute or tag is not found
+  */
+HARDWARE_INTERFACE_PUBLIC
+std::vector<HardwareInfo> parse_control_ressources_from_urdf(const std::string & urdf);
+
+/**
+ * \brief Parse a control ressource from an "ros2_control" tag.
+ *
+ * \param ros2_control_it pointer to ros2_control element with informtions about ressource.
+ * \return robot_control_components::ComponentInfo filled with informations about the robot
+ * \throws std::runtime_error if a attributes or tag are not found
+ */
+HARDWARE_INTERFACE_PUBLIC
+HardwareInfo parse_ressource_from_xml(const tinyxml2::XMLElement * ros2_control_it);
+
+/**
+ * \brief Gets value of the attribute on an XMLelement.
+ * If attribute is not found throws an error.
+ *
+ * \param element_it XMLElement iterator to search for the attribute
+ * \param attribute_name atribute name to serach for and return value
+ * \param tag_name parent tag name where attribute is searched for (used for error output)
+ * \return attribute value
+ * \throws std::runtime_error if attribute is not found
+ */
+HARDWARE_INTERFACE_PUBLIC
+std::string get_attribute_value(
+  const tinyxml2::XMLElement * element_it,
+  const char * attribute_name, const char * tag_name);
+
+/**
+ * \brief Gets value of the attribute on an XMLelement.
+ * If attribute is not found throws an error.
+ *
+ * \param element_it XMLElement iterator to search for the attribute
+ * \param attribute_name atribute name to serach for and return value
+ * \param tag_name parent tag name where attribute is searched for (used for error output)
+ * \return attribute value
+ * \throws std::runtime_error if attribute is not found
+ */
+HARDWARE_INTERFACE_PUBLIC
+std::string get_attribute_value(
+  const tinyxml2::XMLElement * element_it,
+  const char * attribute_name, std::string tag_name);
+
+/**
+  * \brief Search XML snippet from URDF for informations about a control component.
+  *
+  * \param component_it pointer to the iterator where component info should be found
+  * \return robot_control_components::ComponentInfo filled with informations about component
+  * \throws std::runtime_error if a component attribute or tag is not found
+  */
+HARDWARE_INTERFACE_PUBLIC
+ComponentInfo parse_component_from_xml(const tinyxml2::XMLElement * component_it);
+
+/**
+ * @brief Search XML snippet for definition of interfaceTypes.
+ *
+ * @param interfaces_it pointer to the interator over interfaces
+ * @param interfaceTag interface type tag (command or state)
+ * @return std::vector< std::__cxx11::string > list of interface types
+ * \throws std::runtime_error if the interfaceType text not set in a tag
```

Comment:
please stay consistent with your doxygen style

Reply:
Sorry, it is my IDE and it's auto generator....